### DISCUSSION TITLE
Unpublish older deprecated versions of a package

### DISCUSSION THREAD

**Discussion by Irina369** [npm/feedback#785] [General]
Hello!

We had a scenario where we wanted to unpublish a few older, deprecated versions of our package here: https://www.npmjs.com/package/@sap/ux-specification

The versions we wanted to deprecate followed all requirements in the [NPM unpublish policy](https://docs.npmjs.com/policies/unpublish) including only having a few downloads in the last week but because the entire package had more than 300 downloads we were unable to unpublish on our own. It was only with the help of the NPM/GitHub support team that we were able to unpublish these older packages. 

The support team suggested we open a discussion topic to see if we can have a way to unpublish such packages on our own going forward. 

Thank you,
Irina

**Comment by ljharb** [comment]
This would be great; as it is currently, any package which has sufficient downloads can never automatically unpublish an erroneous patch version, even if nobody's downloaded it yet.

**Comment by kroitor** [comment]
We are having issues unpublishing old versions of the package:

```
~/ccxt $ npm unpublish "ccxt@1.40.26" -f
npm WARN using --force Recommended protections disabled.
npm ERR! code E405
npm ERR! 405 Method Not Allowed - PUT https://registry.npmjs.org/ccxt/-rev/13795-5daac910a025be876bf6869f93557a81 - You can no longer unpublish this package.
npm ERR! Failed criteria:
npm ERR! has dependent packages in the registry, has too many downloads
npm ERR!
npm ERR! Please deprecate it instead:
npm ERR! npm deprecate -f 'ccxt@1.40.26' "this package has been deprecated"
npm ERR! To learn more about our unpublish policies, see https://www.npmjs.com/policies/unpublish

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/igorkroitor/.npm/_logs/2023-01-25T16_23_34_824Z-debug.log
~/ccxt $ cat /Users/igorkroitor/.npm/_logs/2023-01-25T16_23_34_824Z-debug.log
0 verbose cli [
0 verbose cli   '/Users/igorkroitor/.nvm/versions/node/v16.13.2/bin/node',
0 verbose cli   '/Users/igorkroitor/.nvm/versions/node/v16.13.2/bin/npm',
0 verbose cli   'unpublish',
0 verbose cli   'ccxt@1.40.26',
0 verbose cli   '-f'
0 verbose cli ]
1 info using npm@8.1.2
2 info using node@v16.13.2
3 timing npm:load:whichnode Completed in 1ms
4 timing config:load:defaults Completed in 0ms
5 timing config:load:file:/Users/igorkroitor/.nvm/versions/node/v16.13.2/lib/node_modules/npm/npmrc Completed in 1ms
6 timing config:load:builtin Completed in 1ms
7 timing config:load:cli Completed in 0ms
8 timing config:load:env Completed in 1ms
9 timing config:load:file:/Users/igorkroitor/ccxt/.npmrc Completed in 0ms
10 timing config:load:project Completed in 0ms
11 timing config:load:file:/Users/igorkroitor/.npmrc Completed in 0ms
12 timing config:load:user Completed in 1ms
13 timing config:load:file:/Users/igorkroitor/.nvm/versions/node/v16.13.2/etc/npmrc Completed in 0ms
14 timing config:load:global Completed in 0ms
15 timing config:load:validate Completed in 0ms
16 timing config:load:credentials Completed in 0ms
17 timing config:load:setEnvs Completed in 0ms
18 timing config:load Completed in 3ms
19 timing npm:load:configload Completed in 3ms
20 timing npm:load:setTitle Completed in 6ms
21 timing npm:load:setupLog Completed in 0ms
22 timing config:load:flatten Completed in 2ms
23 timing npm:load:cleanupLog Completed in 0ms
24 timing npm:load:configScope Completed in 0ms
25 timing npm:load:projectScope Completed in 0ms
26 warn using --force Recommended protections disabled.
27 timing npm:load Completed in 14ms
28 silly unpublish args[0] ccxt@1.40.26
29 silly unpublish spec Result {
29 silly unpublish   type: 'version',
29 silly unpublish   registry: true,
29 silly unpublish   where: undefined,
29 silly unpublish   raw: 'ccxt@1.40.26',
29 silly unpublish   name: 'ccxt',
29 silly unpublish   escapedName: 'ccxt',
29 silly unpublish   scope: undefined,
29 silly unpublish   rawSpec: '1.40.26',
29 silly unpublish   saveSpec: null,
29 silly unpublish   fetchSpec: '1.40.26',
29 silly unpublish   gitRange: undefined,
29 silly unpublish   gitCommittish: undefined,
29 silly unpublish   hosted: undefined
29 silly unpublish }
30 timing command:unpublish Completed in 30823ms
31 verbose stack HttpErrorGeneral: 405 Method Not Allowed - PUT https://registry.npmjs.org/ccxt/-rev/13795-5daac910a025be876bf6869f93557a81 - You can no longer unpublish this package.
31 verbose stack Failed criteria:
31 verbose stack has dependent packages in the registry, has too many downloads
31 verbose stack
31 verbose stack Please deprecate it instead:
31 verbose stack npm deprecate -f 'ccxt@1.40.26' "this package has been deprecated"
31 verbose stack To learn more about our unpublish policies, see https://www.npmjs.com/policies/unpublish
31 verbose stack     at /Users/igorkroitor/.nvm/versions/node/v16.13.2/lib/node_modules/npm/node_modules/npm-registry-fetch/check-response.js:95:15
31 verbose stack     at processTicksAndRejections (node:internal/process/task_queues:96:5)
31 verbose stack     at async unpublish (/Users/igorkroitor/.nvm/versions/node/v16.13.2/lib/node_modules/npm/node_modules/libnpmpublish/unpublish.js:71:7)
31 verbose stack     at async Unpublish.unpublish (/Users/igorkroitor/.nvm/versions/node/v16.13.2/lib/node_modules/npm/lib/unpublish.js:121:9)
32 verbose statusCode 405
33 verbose pkgid ccxt@1.40.26
34 verbose cwd /Users/igorkroitor/ccxt
35 verbose Darwin 21.5.0
36 verbose argv "/Users/igorkroitor/.nvm/versions/node/v16.13.2/bin/node" "/Users/igorkroitor/.nvm/versions/node/v16.13.2/bin/npm" "unpublish" "ccxt@1.40.26" "-f"
37 verbose node v16.13.2
38 verbose npm  v8.1.2
39 error code E405
40 error 405 Method Not Allowed - PUT https://registry.npmjs.org/ccxt/-rev/13795-5daac910a025be876bf6869f93557a81 - You can no longer unpublish this package.
40 error Failed criteria:
40 error has dependent packages in the registry, has too many downloads
40 error
40 error Please deprecate it instead:
40 error npm deprecate -f 'ccxt@1.40.26' "this package has been deprecated"
40 error To learn more about our unpublish policies, see https://www.npmjs.com/policies/unpublish
41 verbose exit 1
```

```
~/ccxt $ npm unpublish "ccxt@1.41.5" -f --verbose
npm verb cli /Users/igorkroitor/.nvm/versions/node/v16.13.2/bin/node /Users/igorkroitor/.nvm/versions/node/v16.13.2/bin/npm
npm info using npm@9.3.1
npm info using node@v16.13.2
npm verb title npm unpublish ccxt@1.41.5
npm verb argv "unpublish" "ccxt@1.41.5" "--force" "--loglevel" "verbose"
npm verb logfile logs-max:10 dir:/Users/igorkroitor/.npm/_logs/2023-01-26T00_29_29_741Z-
npm verb logfile /Users/igorkroitor/.npm/_logs/2023-01-26T00_29_29_741Z-debug-0.log
npm WARN using --force Recommended protections disabled.
npm http fetch GET 200 https://registry.npmjs.com/ccxt?write=true 8333ms (cache updated)
npm http fetch GET 200 https://registry.npmjs.com/ccxt?write=true 5928ms (cache updated)
npm http fetch PUT 404 https://registry.npmjs.com/ccxt/-rev/13798-e4fa07b92887df7e169b29688ab11a0d 12778ms
- ccxt@1.41.5
npm verb exit 0
npm info ok
```

The last HTTP PUT request always ends with a 404 response or a 405 response, regardless of which version I'm trying to unpublish and regardless of which version of npm I'm using. Some of the package versions have 0 downloads, so I assume those should be unpublishable at least. Can anyone shed any light on that? We're stuck, cause we can't publish new versions as we've ran out of npm space quota, and can't unpublish the old versions either (errors shown above).

**Reply by MylesBorins** [reply]
please open a ticket with support https://npmjs.com/support

**Comment by Irina369** [comment]
Hello!

I would like to revisit this request. Is it possible to allow us to **unpublish only specific versions** of a given package given that the version satisfies the [NPM unpublish policy](https://docs.npmjs.com/policies/unpublish) while the overall package does not.

Thank you,
Irina

**Reply by MylesBorins** [reply]
We are not considering the ability to unpublish anything other than [pre-release versions as defined by the semver-spec](https://semver.org/#spec-item-9)... with that said we are still determining if we want to even allow that and do not have a timeline to offering the feature

**Reply by hitesh-parmar** [reply]
Hi @MylesBorins As per the below text in https://docs.npmjs.com/policies/unpublish, it looks like there can be a situation where not all versions of the package could be unpublished. I wonder if such a scenario can exist because as per the npm unpublish policy which talks about the entire package, either all versions are compliant with the npm unpublish policy or none of them are. 

But if certain packages can be compliant at a time, this means we should be able to unpublish certain versions that meet unpublish criteria.
Could you please check this? 

> How to unpublish
> To unpublish a single package version, run npm unpublish <package_name>@<version>.
> If all the versions of a package can be unpublished, you can unpublish all versions at once by running npm unpublish <package_name> --force.

**Reply by MylesBorins** [reply]
The npm registry is treated as immutable. As of today packages cannot unpublish specific versions, as you pointed out.

There are some extreme situations where we will unpublish individual versions of packages such as malware distribution, but in general we do not allow the unpublish of specific versions of a package.

What we are considering is allowing the unpublish of pre release versions, and only pre release versions. Even allowing for this exception is controversial and it is unclear if we are going to move forward with the change.

**Reply by hitesh-parmar** [reply]
Thanks @MylesBorins, 

This means that we have to keep providing security fixes for deprecated version 1.90.x even though they have less than 10 downloads. I hope this gets better.

Additionally, I would propose updating the text 

`To unpublish a single package version, run npm unpublish <package_name>@<version>`
to simply ➡`In extreme situations like malware distribution, you can unpublish specific version of a package by running npm unpublish <package_name>@<version>` 

and 

`If all the versions of a package can be unpublished, you can unpublish all versions at once by running npm unpublish <package_name> --force.` 
to simply ➡ `To unpublish all versions at once, run npm unpublish <package_name> --force.`

**Reply by MylesBorins** [reply]
You cannot unpublish versions yourself, you need to reach out to support and have the team do it. The doc updates you are suggesting are not aligned with the current process, nor do we want to encourage folks to attempt to remove package versions. As such I don't think any doc updates are appropriate today

**Reply by hitesh-parmar** [reply]
Additional question, when you say extreme cases like malware distribution, I guess this means security issues as well, correct? Then we will only request for unpublish of our version if there is a security issue. 

> You cannot unpublish versions yourself, you need to reach out to support and have the team do it.

Do you mean for removing specific versions in case of extreme situations? Yes, this is clear now,  but this is also not reflected in the current documentation. As per docu, support should be contacted in case the package meets unpublish criteria but there are issues with unpublish https://docs.npmjs.com/policies/unpublish#issues 

I am mainly interested in the below change as the usage of `if` implies that there can be a situation when not all packages can be unpublished, but as we discussed as per the policy this situation can never happen.

`If all the versions of a package can be unpublished, you can unpublish all versions at once by running npm unpublish <package_name> --force.` to simply ➡ `To unpublish all versions at once, run npm unpublish <package_name> --force.`

Docu for [Unpublish single version of a package](https://docs.npmjs.com/unpublishing-packages-from-the-registry#unpublishing-a-single-version-of-a-package) also does not mention anything about npm unpublish rules for a specific version.

Here https://blog.npmjs.org/post/190553543620/changes-to-npmunpublish-policy-january-2020 we read "In our experience, 300 downloads over the prior week is the right threshold for safely automating _the removal of a package or package version_ from the registry. "  It is also not clear here that to unpublish a single version, the entire package should have less than 300 downloads.

It is mentioned in several places in the document that a single version of a package can be unpublished but the npm publish policy is only referring to the entire package. This is a gap that should be fixed.

Also here https://docs.npmjs.com/unpublishing-packages-from-the-registry#when-to-deprecate, the text is referring to the whole package, but good that referred link properly refers to package or package version https://docs.npmjs.com/deprecating-and-undeprecating-packages-or-package-versions

**Reply by Irina369** [reply]
Hi Myles,

Is this document meant for collaborators/maintainers of npmjs packages: https://docs.npmjs.com/policies/unpublish 

If so, I think you definitely need to make it very clear that unpublishing is a task that can only be accomplished by your support team and not individuals associated with a package. As it's currently written, under the '[How to unpublish](https://docs.npmjs.com/policies/unpublish#how-to-unpublish)' heading, you are giving directions to end-users on how to unpublish themselves which is confusing.

`To unpublish a single package version, run npm unpublish <package_name>@<version>.`

**Reply by MylesBorins** [reply]
> Is this document meant for collaborators/maintainers of npmjs packages: https://docs.npmjs.com/policies/unpublish

It isn't. If you project meets the unpublish requirements you can unpublish specific versions. If you meet the unpublishing requirements you can unpublish yourself

**Reply by Irina369** [reply]
This is very confusing.

In your previous [comment from yesterday](https://github.com/npm/feedback/discussions/785#discussioncomment-5905728) you say '_**You cannot unpublish versions yourself, you need to reach out to support and have the team do it**_.' but in your [comment today](https://github.com/npm/feedback/discussions/785#discussioncomment-5916201) you say '_**If you project meets the unpublish requirements you can unpublish specific versions. If you meet the unpublishing requirements you can unpublish yourself**_'

So which one is it - can we unpublish ourselves or not? If not, then you definitely need to update the documentation (which is visible and accessible by the end user) to make it very clear who can unpublish - **us (_collaborators/maintainers/users of npmjs_) or you (_npmjs support_)**?

**Reply by MylesBorins** [reply]
The first comment was about a specific scenario where a package has more than 300 downloads + doesn't meet our unpublish guidelines. 

If the package is able to be unpublished as per policy the entire package or individual publishes can be unpublished at any time via the CLI. If the package does not meet the unpublish criteria it cannot be unpublished by an individual, for extreme circumstances they can reach out to support for help

The policy at https://docs.npmjs.com/policies/unpublish applies to packages all up. This includes individual versions. I've asked @monishcm, PM on npm, to review these docs and look for areas for improvement.
